File: sdk-go/abi.go
package range_watch

import (
	"encoding/binary"
	"unsafe"
)

type Notice struct {
	Val uint64
	IDs [][]byte
}

var (
	_bufCap uint32 = 16 << 10 // 16KB
	_bufLen uint32
	_buf           = make([]byte, int(_bufCap))
	_errCap uint32 = 1 << 10 // 1KB
	_errLen uint32
	_val    uint64
	_err    = make([]byte, int(_errCap))
	meta    = make([]uint32, 7)

	recv func(items []Notice)
)

//export __range_watch
func __range_watch() (res uint32) {
	for i, p := range []unsafe.Pointer{
		unsafe.Pointer(&_buf[0]),
		unsafe.Pointer(&_bufCap),
		unsafe.Pointer(&_bufLen),
		unsafe.Pointer(&_err[0]),
		unsafe.Pointer(&_errCap),
		unsafe.Pointer(&_errLen),
		unsafe.Pointer(&_val),
	} {
		meta[i] = uint32(uintptr(p))
	}
	return uint32(uintptr(unsafe.Pointer(&meta[0])))
}

var notices []Notice

//export __range_watch_recv
func __range_watch_recv() {
	notices = notices[:0]
	var i uint32 = 2
	count := binary.BigEndian.Uint16(_buf)
	for range count {
		var n Notice
		n.Val = binary.BigEndian.Uint64(_buf[i:])
		i += 8
		idCount := binary.BigEndian.Uint16(_buf[i:])
		i += 2
		for range idCount {
			idLen := uint32(binary.BigEndian.Uint16(_buf[i:]))
			i += 2
			id := _buf[i : i+idLen]
			i += idLen
			n.IDs = append(n.IDs, id)
		}
		notices = append(notices, n)
	}
	recv(notices)
}

func setData(b []byte) {
	_bufLen = uint32(len(b))
	copy(_buf[:len(b)], b)
}

func getErr() (e error) {
	if _errLen > 0 {
		e = strErr(string(_err[:_errLen]))
	}
	return
}

func appendKey(k []byte) bool {
	if _bufLen+2+uint32(len(k)) > _bufCap {
		return false
	}
	binary.BigEndian.PutUint16(_buf[_bufLen:], uint16(len(k)))
	_bufLen += 2
	copy(_buf[_bufLen:], k)
	_bufLen += uint32(len(k))
	return true
}

//go:wasm-module pantopic/wazero-range-watch
//export __range_watch_reserve
func _reserve()

//go:wasm-module pantopic/wazero-range-watch
//export __range_watch_open
func _open()

//go:wasm-module pantopic/wazero-range-watch
//export __range_watch_start
func _start()

//go:wasm-module pantopic/wazero-range-watch
//export __range_watch_stop
func _stop()

//go:wasm-module pantopic/wazero-range-watch
//export __range_watch_queue
func _queue()

//go:wasm-module pantopic/wazero-range-watch
//export __range_watch_flush
func _flush()

//go:wasm-module pantopic/wazero-range-watch
//export __range_watch_clear
func _clear()

//go:wasm-module pantopic/wazero-range-watch
//export __range_watch_group_start
func _group_start()

//go:wasm-module pantopic/wazero-range-watch
//export __range_watch_group_stop
func _group_stop()

// Fix for lint rule `unusedfunc`
var _ = __range_watch
var _ = __range_watch_recv

File: sdk-zig/src/abi.zig
//! Zig SDK for the pantopic/wazero-range-watch host module.
//! Mirrors the ABI of sdk-go: the guest exports `__range_watch` (meta page)
//! and `__range_watch_recv`, and imports the host functions below.

const std = @import("std");
const sdk = @import("sdk.zig");

const _buf_cap_default: u32 = 16 << 10; // 16KB
const _err_cap_default: u32 = 1 << 10; // 1KB
const _vals_cap_default: u32 = 1 << 10; // 1,024

pub var _buf_cap: u32 = _buf_cap_default;
pub var _buf_len: u32 = 0;
pub var _buf: [_buf_cap_default]u8 = undefined;
pub var _err_cap: u32 = _err_cap_default;
pub var _err_len: u32 = 0;
pub var _err: [_err_cap_default]u8 = undefined;
pub var _val: u64 = 0;
pub var _meta: [7]u32 = undefined;

pub var _recv: ?*const fn (notices: []Notice) void = null;
pub var _allocator: std.mem.Allocator = undefined;

export fn __range_watch() u32 {
    _meta = .{
        @intFromPtr(&_buf),
        @intFromPtr(&_buf_cap),
        @intFromPtr(&_buf_len),
        @intFromPtr(&_err),
        @intFromPtr(&_err_cap),
        @intFromPtr(&_err_len),
        @intFromPtr(&_val),
    };
    return @intFromPtr(&_meta);
}

pub const Notice = struct {
    val: u64,
    ids: [][]const u8,
};

export fn __range_watch_recv() void {
    var i: u32 = 0;
    const count = std.mem.readInt(u16, @ptrCast(&_buf[i]), .big);
    const notices = _allocator.alloc(Notice, count) catch unreachable;
    i += 2;
    for (notices) |*n| {
        n.* = Notice{
            .val = std.mem.readInt(u64, @ptrCast(&_buf[i]), .big),
            .ids = _allocator.alloc([]const u8, std.mem.readInt(u16, @ptrCast(&_buf[i + 8]), .big)) catch unreachable,
        };
        i += 10;
        for (n.ids) |*id| {
            const len = std.mem.readInt(u16, @ptrCast(&_buf[i]), .big);
            i += 2;
            id.* = _buf[i .. i + len];
            i += len;
        }
    }
    if (_recv) |f| f(notices);
}

pub extern "pantopic/wazero-range-watch" fn __range_watch_queue() void;
pub extern "pantopic/wazero-range-watch" fn __range_watch_flush() void;
pub extern "pantopic/wazero-range-watch" fn __range_watch_reset() void;
pub extern "pantopic/wazero-range-watch" fn __range_watch_reserve() void;
pub extern "pantopic/wazero-range-watch" fn __range_watch_open() void;
pub extern "pantopic/wazero-range-watch" fn __range_watch_start() void;
pub extern "pantopic/wazero-range-watch" fn __range_watch_stop() void;
pub extern "pantopic/wazero-range-watch" fn __range_watch_group_start() void;
pub extern "pantopic/wazero-range-watch" fn __range_watch_group_stop() void;

pub fn setData(b: []const u8) void {
    _buf_len = @intCast(b.len);
    @memcpy(_buf[0..b.len], b);
}

pub fn getErr() sdk.Error!void {
    if (_err_len > 0) return sdk.Error.Host;
}

pub fn appendKey(k: []const u8) bool {
    if (_buf_len + 2 + k.len > _buf_cap) {
        return false;
    }
    std.mem.writeInt(u16, _buf[_buf_len..][0..2], @intCast(k.len), .big);
    _buf_len += 2;
    @memcpy(_buf[_buf_len..][0..k.len], k);
    _buf_len += @intCast(k.len);
    return true;
}
